org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/PomFacadeAdapterFactory.java

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *      Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.m2e.core.internal.project;
+
+import java.util.Optional;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IAdapterFactory;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+
+import org.eclipse.m2e.core.embedder.IPomFacade;
+import org.eclipse.m2e.core.embedder.MavenModelManager;
+import org.eclipse.m2e.core.internal.IMavenConstants;
+import org.eclipse.m2e.core.project.IMavenProjectFacade;
+import org.eclipse.m2e.core.project.IMavenProjectRegistry;
+
+
+@Component(service = IAdapterFactory.class, property = {
+    IAdapterFactory.SERVICE_PROPERTY_ADAPTABLE_CLASS + "=org.eclipse.core.resources.IResource",
+    IAdapterFactory.SERVICE_PROPERTY_ADAPTER_NAMES + "=org.eclipse.m2e.core.embedder.IPomFacade"})
+public class PomFacadeAdapterFactory implements IAdapterFactory {
+
+  @Reference
+  private IMavenProjectRegistry mavenProjectRegistry;
+
+  @Reference
+  private MavenModelManager mavenModelManager;
+
+  @Override
+  public <T> T getAdapter(Object adaptableObject, Class<T> adapterType) {
+    if(adapterType == IPomFacade.class) {
+      if(adaptableObject instanceof IFile) {
+        IFile file = (IFile) adaptableObject;
+        if(file.getName().equalsIgnoreCase(IMavenConstants.POM_FILE_NAME)) {
+          return adapterType.cast(getFacadeForPom(file));
+        }
+        //TODO not a pom file, but actually pom files can have any name...
```

Comment:
Maybe it should be looked into the file if it defines the corresponding XML header and starts with a suitable element to decide if it is good to go further. Or does the look-up fail gracefully if it is an unrelated file.

Alternatively the container could be used in this case. But then the 'Run as Maven build' would be available for all files in a (pomless) Module. This is probably not desired.

Reply:
Using the container is bad as one can't decide then if this is really something that could be executed as maven, this should be avoided. Beside that m2e currently do not support other things than pom elsewhere so we should not add more magic here.

Reply:
Sound's reasonable. These were just some thought I wanted to put on the table.